File: priv/combined_metrics/samples/code_smells/no_fixme_comments/good/importer.js
// Handles importing data from CSV and external sources.

const fs = require("fs");

const EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

function importCsv(path) {
  const content = fs.readFileSync(path, "utf8");
  if (content === "") {
    return { error: "empty_file" };
  }

  const rows = content
    .split("\n")
    .filter((line) => line.trim() !== "")
    .map(parseRow)
    .filter((row) => row !== null);

  return { ok: rows };
}

function parseRow(line) {
  const parts = line.split(",");
  if (parts.length !== 3) {
    return null;
  }
  const [id, name, email] = parts;
  return { id, name, email };
}

function importUsers(rows) {
  const results = rows.map(insertUser);
  const ok = results.filter((r) => r.ok).length;
  const errors = results.length - ok;
  return { ok, errors };
}

function validateRow(row) {
  const normalized = row.email.toLowerCase();
  if (EMAIL_REGEX.test(normalized)) {
    return { ok: { ...row, email: normalized } };
  }
  return { error: "invalid_email" };
}

function deduplicate(rows) {
  const seen = new Map();
  for (const row of rows) {
    const key = row.email.toLowerCase();
    if (!seen.has(key)) {
      seen.set(key, { ...row, email: key });
    }
  }
  return Array.from(seen.values());
}

function importFromApi(sourceUrl) {
  const fetched = fetchRemote(sourceUrl);
  if (fetched.error) {
    return fetched;
  }
  return parseApiResponse(fetched.ok);
}

function transformRow(row, fieldMap) {
  const result = {};
  for (const [src, dst] of Object.entries(fieldMap)) {
    result[dst] = row[src];
  }
  return result;
}

function writeResults(results, outputPath) {
  const backupPath = `${outputPath}.bak`;
  if (fs.existsSync(outputPath)) {
    fs.copyFileSync(outputPath, backupPath);
  }
  const content = results.map(formatResult).join("\n");
  fs.writeFileSync(outputPath, content);
  return { ok: true };
}

function insertUser(row) {
  return { ok: row };
}

function fetchRemote(_url) {
  return { ok: [] };
}

function parseApiResponse(data) {
  return { ok: data };
}

function formatResult(result) {
  return JSON.stringify(result);
}

module.exports = {
  importCsv,
  parseRow,
  importUsers,
  validateRow,
  deduplicate,
  importFromApi,
  transformRow,
  writeResults,
};

File: priv/combined_metrics/samples/code_smells/no_fixme_comments/good/importer.py
"""Handles importing data from CSV and external sources."""

import re
from pathlib import Path


EMAIL_REGEX = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")


def import_csv(path):
    text = Path(path).read_text()
    if not text:
        return {"error": "empty_file"}

    rows = []
    for line in text.split("\n"):
        if not line.strip():
            continue
        parsed = parse_row(line)
        if parsed is not None:
            rows.append(parsed)
    return {"ok": rows}


def parse_row(line):
    parts = line.split(",")
    if len(parts) != 3:
        return None
    row_id, name, email = parts
    return {"id": row_id, "name": name, "email": email}


def import_users(rows):
    ok, errors = [], []
    for row in rows:
        result = insert_user(row)
        if result.get("ok"):
            ok.append(result)
        else:
            errors.append(result)
    return {"ok": len(ok), "errors": len(errors)}


def validate_row(row):
    normalized = row["email"].lower()
    if EMAIL_REGEX.match(normalized):
        return {"ok": {**row, "email": normalized}}
    return {"error": "invalid_email"}


def deduplicate(rows):
    seen = {}
    for row in rows:
        key = row["email"].lower()
        if key not in seen:
            seen[key] = {**row, "email": key}
    return list(seen.values())


def import_from_api(source_url):
    fetched = fetch_remote(source_url)
    if "error" in fetched:
        return fetched
    return parse_api_response(fetched["ok"])


def transform_row(row, field_map):
    return {dst: row.get(src) for src, dst in field_map.items()}


def write_results(results, output_path):
    output = Path(output_path)
    backup = output.with_suffix(output.suffix + ".bak")

    if output.exists():
        backup.write_bytes(output.read_bytes())

    content = "\n".join(format_result(r) for r in results)
    output.write_text(content)
    return {"ok": True}


def insert_user(row):
    return {"ok": row}


def fetch_remote(_url):
    return {"ok": []}


def parse_api_response(data):
    return {"ok": data}


def format_result(result):
    return repr(result)
